autopush/endpoint.py

```diff
@@ -220,3 +256,115 @@ def write_error(self, code, exception=None):
         if exception is not None:
             log.err(exception)
         self.finish()
+
+
+class RegistrationHandler(cyclone.web.RequestHandler):
+
+    def _error_response(self, failure):
+        log.err(failure)
+        self.set_status(500)
+        self.write("Error processing request")
+        self.finish()
+
+    def _load_params(self):
+        tags = {'chid': 'channelid',
+                'conn': 'connect',
+                }
+        chid = conn = None
+        if len(self.request.body) > 0:
+            body = urlparse.parse_qs(self.request.body, keep_blank_values=True)
+            chid = body.get(tags['chid'], [None])[0]
+            conn = body.get(tags['conn'], [None])[0]
+        if chid is None:
+            chid = self.request.arguments.get(tags['chid'], [None])[0]
+        if conn is None:
+            conn = self.request.arguments.get(tags['conn'], [None])[0]
+
+        if conn is None:
+            log.msg("Missing conn %s" % (conn))
+            return False
+
+        if chid is None or len(chid) == 0:
+            chid = str(uuid.uuid4())
+
+        self.chid = chid
+        self.conn = conn
+        return True
+
+    def initialize(self):
+        self.metrics = self.ap_settings.metrics
+
+    def options(self, token):
+        self._addCors()
+
+    def head(self, token):
+        self._addCors()
+
+    def _addCors(self):
+        if self.ap_settings.cors:
+            self.set_header("Access-Control-Request-Method", "*")
+
+    def _error(self, code, msg):
+        self.set_status(code, msg)
+        self.finish()
+        return
+
+    @cyclone.web.asynchronous
+    def get(self, uaid=None):
+        if uaid is None:
+            return self._error(400, "invalid UAID")
+        try:
+            uuid.UUID(uaid)
+        except Exception, e:
+            log.msg("Improper UAID value specified %s" % e)
+            return self._error(400, "invalid UAID")
+        self.uaid = uaid
+
+        self.chid = str(uuid.uuid4())
+        self._registered(True)
+
+    @cyclone.web.asynchronous
+    def put(self, uaid=None):
+        self.metrics = self.ap_settings.metrics
+        self.start_time = time.time()
+
+        self.add_header("Content-Type", "application/json")
+
+        if uaid is None:
```

Comment:
What if we included the ID in the body, instead of the URL—and just made the URL `/register`? It seems a bit odd to make a request to `/register/abc` but receive a response with `"useragentid": "xyz"`.